discards, deconstruction, and assignment (#1730)
dotnet/csharpstandard: standard/statements.md

#### 13.9.5.4 Deconstructing foreach

> A deconstructing foreach replaces the declaration and initialisation of a single iteration variable per iteration, in the synchronous and asynchronous foreach statements, with a *deconstructor* ([§12.23.3](expressions.md#12233-deconstructing-assignment)) whose *deconstructor_element*s are matched, per iteration, against the elements obtained by deconstructing each collection element.

Reviewer: In C# (but not all languages) deconstruction is distinct from pattern matching. The former selects values from a compound value and cannot fail, the latter *matches* against a value and can fail.

So this change as written is wrong, there is no matching involved. This either needs rewriting or simply reverting:

```suggestion
A deconstructing foreach replaces the declaration and initialisation of a single iteration variable per iteration, in the synchronous and asynchronous foreach statements, with a collection of zero or more iteration variables declared per iteration within the *deconstructor*([§12.23.3](expressions.md#12233-deconstructing-assignment)) of a *deconstructing_assignment*.
```

The “curious” semantic here is that “zero or more” iteration variables are allowed. Given the semantics are partly reconstructed based on an implementation it should be determined whether this is an “accident of implementation” and the number should really be “one or more”.

Author: I accepted the suggestion, and then unresolved to make sure I remember to address the curious semantic here.

> Each *deconstructor_element* shall be either a *declaration_expression*, which declares an iteration variable, or a discard, which is a placeholder that matches its corresponding element without declaring a variable. It is a compile-time error for any *deconstructor_element* to be a *variable_reference*.

Reviewer: See previous comment on line 1444, there is no matching. Same action required: either reword or revert:

```suggestion
All variables assigned to by the *deconstructor* must be declared within the *deconstructor*, it is a compile time error for any *deconstructor_element* to be a *variable_reference*.```

}
```

> This follows the behavior of synchronous foreach ([§13.9.5.2](statements.md#13952-synchronous-foreach)), differing by replacing the declaration and initialisation of a single iteration variable with a *deconstructing_assignment* whose *deconstructor* declares an iteration variable for each *declaration_expression* it contains (discards declare none):

Reviewer: This change doesn’t read well to me – it is saying each *declaration_expression" declares a variable…

I suggest either rewriting or reverting as for the above cases, but do not forget the “curious” semantic mentioned for 1444 above and maybe “revert” to “one or more”.

- `C` and `E` are determined as for synchronous foreach

> - `e` is not visible or accessible anywhere in the program except as indicated in the above code

Reviewer: Egg-on-face ☹️ (IIRC I typed that!) – thanks for fixing.

}
```

> This follows the behavior of asynchronous foreach ([§13.9.5.3](statements.md#13953-asynchronous-foreach)), differing by replacing the declaration and initialisation of a single iteration variable with a *deconstructing_assignment* whose *deconstructor* declares an iteration variable for each *declaration_expression* it contains (discards declare none):

Reviewer: If you are reverting/rewriting following the preceding suggestions do the same here.

> - `enumerator` is not visible or accessible anywhere in the program except as indicated in the above code

Reviewer: Not only did I probably author this on 1477 I copied & pasted it! Jeepers 😫.
